pkg/controllers/metrics/state/node.go

```diff
@@ -0,0 +1,255 @@
+		},
+		nodeLabelNames(),
+	)
+
+	wellKnownLabels = getWellKnownLabels()
+)
+
+func newNodeGaugeVec(name, help string) *prometheus.GaugeVec {
+	return prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      name,
+			Help:      help,
+		},
+		nodeLabelNames(),
+	)
+}
+
+func nodeLabelNames() []string {
+	labels := []string{
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodePhase, // NOTE: deprecated
+	}
+
+	for _, l := range wellKnownLabels {
+		labels = append(labels, l)
+	}
+
+	return labels
+}
+
+func init() {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		metrics.Registry.MustRegister(gauge)
+	}
+}
+
+type NodeScraper struct {
+	cluster  *state.Cluster
+	labelMap map[string]map[*prometheus.GaugeVec][]prometheus.Labels
+}
+
+func NewNodeScraper(cluster *state.Cluster) *NodeScraper {
+	return &NodeScraper{
+		cluster:  cluster,
+		labelMap: make(map[string]map[*prometheus.GaugeVec][]prometheus.Labels),
+	}
+}
+
+func (ns *NodeScraper) Scrape(ctx context.Context) {
+	nodes := sets.NewString()
+	ns.cluster.ForEachNode(func(n *state.Node) bool {
+		if _, ok := ns.labelMap[n.Node.Name]; !ok {
+			ns.labelMap[n.Node.Name] = make(map[*prometheus.GaugeVec][]prometheus.Labels)
+		}
+		nodes.Insert(n.Node.Name)
+
+		// Populate  metrics
+		for gaugeVec, resourceList := range map[*prometheus.GaugeVec]v1.ResourceList{
+			overheadGaugeVec:       ns.getSystemOverhead(n.Node),
+			podRequestsGaugeVec:    resources.Subtract(n.PodTotalRequests, n.DaemonSetRequested),
+			podLimitsGaugeVec:      resources.Subtract(n.PodTotalLimits, n.DaemonSetLimits),
+			daemonRequestsGaugeVec: n.DaemonSetRequested,
+			daemonLimitsGaugeVec:   n.DaemonSetLimits,
+			allocatableGaugeVec:    n.Node.Status.Allocatable,
+		} {
+			ns.set(resourceList, n.Node, gaugeVec)
+		}
+
+		return true
+	})
+
+	ns.cleanup(ctx, nodes)
+}
+
+func (ns *NodeScraper) cleanup(ctx context.Context, existingNodes sets.String) {
+	for node := range sets.NewString(lo.Keys(ns.labelMap)...).Difference(existingNodes) {
+		for gaugeVec, labelSet := range ns.labelMap[node] {
+			for _, labels := range labelSet {
+				gaugeVec.Delete(labels)
+			}
+		}
+		delete(ns.labelMap, node)
+	}
+}
+
+// set sets the value for the node gauge
+func (ns *NodeScraper) set(resourceList v1.ResourceList, node *v1.Node, gaugeVec *prometheus.GaugeVec) {
+	for resourceName, quantity := range resourceList {
+		// Reformat resource type to be consistent with Prometheus naming conventions (snake_case)
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+
+		labels := ns.getNodeLabels(node, resourceTypeName)
+		ns.labelMap[node.Name][gaugeVec] = append(ns.labelMap[node.Name][gaugeVec], labels)
```

Comment:
I'm a bit confused. Doesn't this cause the gaugeVec list to continually grow?

Reply:
You're right this should be modeled as a set rather than a slice.